### ISSUE TITLE
Setting the period: infinity prevents randomized slide order of shuffle: true

### ISSUE THREAD

**Issue by nigef** [justinmc/jQuery-Open-Carousel#2]
Great stuff Justin,

Just wondering if anyone has experienced issue with the shuffling when setting the period to infinity.

I tried this on the unadulterated code and the shuffle stopped working for me.

Thanks,

/n

**Comment by justinmc** [comment]
Definitely could be, I fixed another problem that period =Infinity was causing not too long ago. I'm traveling at the moment, but I'll take a look when I get home in a few days.

**Comment by nigef** [comment]
I tested the shuffle feature with a series of 144 numbers and it has only returned some of the numbers, over many, many instances. It seems only the first 16 show.

**Comment by justinmc** [comment]
I created a jsfiddle for a very simple period=Infinity and shuffle=true carousel with 32 slides, and it seems to work there at least: http://jsfiddle.net/justinmc/YXXnc/.  Can you reproduce the bug in that fiddle?

Just to clarify, the shuffle feature shuffles the order of the slides on initialization, so in the fiddle you can see it reshuffle when you press Run.  It doesn't jump to a random slide when you press the next button or anything like that.  It definitely shouldn't be making some slides inaccessible in any case though...  Are you using any other features like perscroll?

**Comment by nigef** [comment]
I updated fiddle to basic of what I plan on doing and it seems that numbers only get up to 4 x 7.

http://jsfiddle.net/ad3qX/

**Comment by justinmc** [comment]
Now I see what you mean and yes that looks like a bug.  I'm headed to work now so I'll take a crack at this when I get back and hopefully have a fix within the day.

Thanks for pointing this out, and really cool use as flashcards by the way...

**Comment by nigef** [comment]
Thanks I really hope it can be solved because I look forward to using.

**Comment by justinmc** [comment]
Looks like it's fixed, check out the updated fiddle: http://jsfiddle.net/justinmc/ad3qX/2/.  The short answer is to simply increase the width of the .ocarousel_window_slides div in CSS.  I set it to 10000px originally thinking that would be large enough, but your 100+ slides were reaching the end and wrapping.

I added a zero and changed the width in the repository to 100000px, which is way more than enough for the fiddle's application.  You can git pull to get the changes (SHA: 02e29c4f89d6f48d447ec11444e2e954b41dc7e9), or just add some width to the .ocarousel_window_slides div in your project.

100000px is pretty wide, but I worry about performance problems if I were to make it enormously wide.  If you run into problems like this again, just increase the width a bit more.

As far as I can tell, I cannot make that div expand to fit its contents in CSS, I must set a fixed width.  Maybe setting the width in javascript or a slightly new approach for moving slides around is a more robust solution.  I'll keep that in mind for when I have some time, but for now this solution works fine for this.

**Comment by nigef** [comment]
Works in Chrome browser, but try setting shuffle:true and running in Safari. http://jsfiddle.net/d8G7s/

Shuffle doesn't seem to work as it tends to stay around 12, 1, 2, or 3 range.

Still a bit buggy in Safari.

**Comment by justinmc** [comment]
Well all of your slides are appearing, it's just that they're not shuffled so well, so don't worry about anything being missing.  The only real problem is that my shuffle algorithm sucks (especially in Safari for whatever reason).

I've noticed it didn't seem very thorough before, so I'll try swapping in something more robust right now and make a new commit.

**Comment by nigef** [comment]
If you're changing the shuffle algorithm, while you're at it maybe see if it's worth while to replace the 100,000px width to a z-index solution stacking slides. This would be useful for people doing full-screen slide widths and larger display sizes etc...

**Comment by justinmc** [comment]
I've been thinking about doing stacked slides, but that would get more complicated for situations with multiple slides visible at one time, or slides that are partially visible, like the photo example on the demo page: http://justinmccandless.com/demos/jQuery-Open-Carousel/index.html

Sill should be doable though, maybe I'll take a look.  The thought of a 100,000px div just makes me think I should be doing it in a better way.

**Comment by justinmc** [comment]
Ok I found a fancy algorithm called Fisher Yates that seems to work much better, just commited here SHA: fa4195dc624004c8ea81eff21ad0c455a9cd1d6b and updated the fiddle here: http://jsfiddle.net/justinmc/zPJxK/

**Comment by nigef** [comment]
Okay Justin here is a new challenge for you that I was working on if you're interested.

A button to add and remove a class put a display:none has been added to some of the carousel divs.

When divs are removed from the DOM, they are still seen by the left-right arrows. When the divs are hidden, clicking on the arrow does nothing until it gets past the hidden divs. What can be changed so that the left-right arrows do not see and include the hidden div?

See example here http://codepen.io/nigelNSF/pen/ngEhI

**Comment by justinmc** [comment]
That's an interesting expansion to the project, it's giving me a lot of ideas...

Unfortunately I probably won't have time to take a good look at this until the weekend, but in case you need a quick working solution I've hacked something quick together.  I created a new branch called mutableCarousel, and I edited the functions that find the next/previous divs to scroll to.  They will now skip over any divs that are hidden.  Here's the commit SHA: a6c4b0fcd30a01583c780ef0a59446aea58eaf3a

It doesn't work perfectly with your codepen example right now, but it's a good quick start.  What I should really do is a more robust change data -> rerender model, which would make updating the slide count and things like that very easy.  Expect me to take a better look at that this weekend.

**Comment by nigef** [comment]
Cool, another thing that seems intuitive is to add keyboard inputs to left right function

**Comment by justinmc** [comment]
I just really improved the public interface to Ocarousel and merged my mutableCarousel branch, it should make everything you want to do much easier.

Check out the example with additions and removals at the bottom of the demo now: http://justinmccandless.com/demos/jQuery-Open-Carousel/index.html

I added a whole section in the README about accessing the public methods of the Ocarousel object.  You can access methods like add() and remove() that will edit the carousel and rerender it for you, or even access the data directly yourself.

Take a look at the example and the README and let me know how that helps your situation.

**Comment by justinmc** [comment]
Also for left/right keys scrolling the carousel, I'm not going to implement that directly in the main project because it could cause conflicts when having multiple carousels on the same page, which I want to avoid.  You could pretty easily implement this yourself by setting keydown events that scroll to the next/previous slides, see that programmatic example again.  Let me know if you need help.

**Comment by justinmc** [comment]
Lastly, thinking about stacking slides instead of having a very wide div, I took a look at some widely used websites that also implement similar carousels.  Actually all that I looked at used the same wide div method that I'm using (play.google.com, steampowered.com).  This makes me more relaxed about losing performance because of a very wide div, so I'm going to keep it as it is for now, as long as I don't have a good reason to make the complicated hack to do stacked slides.
